Modules/Filtering/Convolution/include/itkFFTConvolutionImageFilter.hxx

```diff
@@ -126,47 +147,154 @@ FFTConvolutionImageFilter<TInputImage, TKernelImage, TOutputImage, TInternalPrec
   ProgressAccumulator *      progress,
   float                      progressWeight)
 {
-  // Pad the image
-  InputSizeType   padSize = this->GetPadSize();
-  InputRegionType inputRegion = input->GetLargestPossibleRegion();
-  InputSizeType   inputSize = inputRegion.GetSize();
+  float remainingProgress = 1.0f;
+
+  InputRegionType  inputLargestRegion = input->GetLargestPossibleRegion();
+  InputSizeType    inputLargestSize = inputLargestRegion.GetSize();
+  InputIndexType   inputLargestIndex = inputLargestRegion.GetIndex();
+  InputRegionType  inputRequestedRegion = input->GetRequestedRegion();
+  InputSizeType    inputRequestedSize = inputRequestedRegion.GetSize();
+  InputIndexType   inputRequestedIndex = inputRequestedRegion.GetIndex();
+  OutputRegionType outputRequestedRegion = this->GetOutput()->GetRequestedRegion();
+  OutputSizeType   outputRequestedSize = outputRequestedRegion.GetSize();
+  OutputIndexType  outputRequestedIndex = outputRequestedRegion.GetIndex();
+
+  // Pad the input image such that the requested region, expanded by
+  // twice the kernel radius, lies entirely within the buffered region.
+  // If the requested region is at least the kernel radius away from
+  // all boundaries of the buffered input region then padding is skipped.
 
   using InputPadFilterType = PadImageFilter<InputImageType, InputImageType>;
-  auto inputPadder = InputPadFilterType::New();
-  inputPadder->SetBoundaryCondition(this->GetBoundaryCondition());
+  using PadSizeType = typename InputPadFilterType::SizeType;
 
-  InputSizeType inputLowerBound = this->GetPadLowerBound();
-  inputPadder->SetPadLowerBound(inputLowerBound);
+  PadSizeType    lowerPad, upperPad;
+  bool           needsKernelPadding = false;
+  KernelSizeType kernelRadius = this->GetKernelRadius();
+  for (unsigned int dim = 0; dim < ImageDimension; ++dim)
+  {
+    signed int lower = static_cast<signed int>(kernelRadius[dim]) - (inputRequestedIndex[dim] - inputLargestIndex[dim]);
+    signed int upper =
+      static_cast<signed int>(kernelRadius[dim]) -
+      ((inputLargestIndex[dim] + inputLargestSize[dim]) - (inputRequestedIndex[dim] + inputRequestedSize[dim]));
```

Comment:
On these lines and just above, my compiler says that an `int` plus an `unsigned int` is an `unsigned int`.  It looks like this code is instead depending upon that sum being an `int`.  However, the final assignment maybe reinterprets what has sometimes accidentally become a very large `unsigned int` back to what it should have been, a negative `int`; this should work but seems unsatisfactory.  Instead maybe declare the `*Size` variables and the `kernelRadius` variable as `InputIndexType`, by casting the right-hand side of their initializations if necessary.

Reply:
Good catch!

Reply:
After review I found that there is not a cast defined for casting from `InputSizeType` to `InputIndexType` as one is an `itk::Size` and the other is an `itk::Offset`. Attempting to cast directly returns a compiler failure. Instead I added two extra lines above to implicitly cast addition operations to the `int` type, which I think also helps with readability.

Reply:
One cast to `int` but not to `itk::Offset` (which is implemented at `int`)?

Reply:
@Leengit I'm not sure I understand the question?